x-pack/agent/pkg/agent/stateresolver/stateresolver.go

```diff
@@ -29,21 +32,32 @@ func NewStateResolver(log *logger.Logger) (*StateResolver, error) {
 }
 
 // Resolve resolves passed config into one or multiple steps
-func (s *StateResolver) Resolve(cfg configrequest.Request) (uid.ID, []configrequest.Step, error) {
+func (s *StateResolver) Resolve(
+	cfg configrequest.Request,
+) (uid.ID, []configrequest.Step, Acker, error) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
 	newState, steps := converge(s.curState, cfg)
 	id, err := uid.Generate()
 	if err != nil {
-		return id, nil, err
+		return id, nil, nil, err
 	}
 
 	s.l.Infof("New State ID is %s", newState.ShortID())
 	s.l.Infof("Converging state requires execution of %d step(s)", len(steps))
 
-	// Keep the should state for next tick.
-	s.curState = newState
+	// Allow the operator to ack the should state when applying the steps is done correctly.
+	ack := func() {
+		s.ack(newState)
+	}
+
+	return id, steps, ack, nil
+}
 
-	return id, steps, nil
+func (s *StateResolver) ack(newState state) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.l.Info("Updating internal state")
```

Comment:
is it possible that these method is run concurrently? e.g two handle calls of operator one slow another fast which would cause faster call to be overwritten by older but slower call?

Reply:
I've looked at the operator and the steps are executed sequentially https://github.com/elastic/beats/blob/fleet/x-pack/agent/pkg/agent/operation/operator.go#L121-L131 So we could remove the mutex from the sate resolver. I believe it was needed when the state resolver was not kept by the operator. So I think we could remove all the locks. WDYT?

Reply:
i think it's fine. if we keep usage only in operator and keep warning somewhere so we dont forget it in case we will rewrite operator to be async or use resolver somewhere else